src/TorchSharp/Utils/PyTorchPickler.cs

```diff
@@ -0,0 +1,217 @@
+            {
+                _archive = archive;
+                _tensorCount = 0;
+            }
+
+            protected override bool persistentId(object pid, out object newpid)
+            {
+                if (pid is not torch.Tensor) {
+                    newpid = null;
+                    return false;
+                }
+
+                var tensor = (torch.Tensor)pid;
+
+                bool copied = false;
+                if (tensor.device_type != DeviceType.CPU) {
+                    tensor = tensor.to(torch.CPU);
+                    copied = true;
+                }
+
+                // The persistentId function in pickler is a way of serializing an object using a different
+                // stream and then pickling just a key representing it.
+                // the data yourself from another source. The `torch.load` function uses this functionality
+                // and lists for the pid a tuple with the following items:
+                // ("storage", storage_type=classDict (e.g., torch.LongTensor), key, location, numElements
+
+                // Start by serializing the object to a file in the archive
+                var entry = _archive.CreateEntry($"model/data/{_tensorCount}");
+                using (var stream = entry.Open())
+                    stream.Write(tensor.bytes.ToArray(), 0, tensor.bytes.Length);
+
+                // Start collecting the items
+                newpid = new object[] {
+                    "storage",
+                    new Storage(GetStorageNameFromScalarType(tensor.dtype)), // storage_type
+                    _tensorCount.ToString(), // key
+                    "cpu", // location
+                    tensor.NumberOfElements // numel
+                };
+
+                // Post-cleanup items
+                _tensorCount++;
+                if (copied) tensor.Dispose();
+
+                return true;
+            }
+
+            static string GetStorageNameFromScalarType(torch.ScalarType storage)
+            {
+                return storage switch {
+                    torch.ScalarType.Float64 => "DoubleStorage",
+                    torch.ScalarType.Float32 => "FloatStorage",
+                    torch.ScalarType.Float16 => "HalfStorage",
+                    torch.ScalarType.Int64 => "LongStorage",
+                    torch.ScalarType.Int32 => "IntStorage",
+                    torch.ScalarType.Int16 => "ShortStorage",
+                    torch.ScalarType.Int8 => "CharStorage",
+                    torch.ScalarType.Byte => "ByteStorage",
+                    torch.ScalarType.Bool => "BoolStorage",
+                    torch.ScalarType.BFloat16 => "BFloat16Storage",
+                    torch.ScalarType.ComplexFloat64 => "ComplexDoubleStorage",
+                    torch.ScalarType.ComplexFloat32 => "ComplexFloatStorage",
+                    _ => throw new NotImplementedException()
+                };
+            }
+        }
+        class Storage
```

Comment:
Even though many of these classes are private, it would be useful to have comments explaining their role. We don't systematically do that in this repo, but it would be useful.

Reply:
Of course, I was working on that now :)